crates/context/config/src/client/env/proxy/types/starknet.rs

```diff
@@ -397,3 +446,143 @@ impl From<StarknetProposalId> for ProposalId {
+        let num_entries = match iter.next() {
+            Some(felt) => felt.to_bytes_be()[31] as usize,
+            None => return Ok(Self { entries: vec![] }),
+        };
+
+        let mut entries = Vec::new();
+
+        // Read exactly num_entries pairs
+        for _ in 0..num_entries {
+            // Get key array length and contents
+            if let Some(key_len) = iter.next() {
+                let key_len = key_len.to_bytes_be()[31] as usize;
+                let mut key = Vec::new();
+                for _ in 0..key_len {
+                    if let Some(felt) = iter.next() {
+                        key.push(*felt);
+                    }
+                }
+
+                // Get value array length and contents
+                if let Some(value_len) = iter.next() {
+                    let value_len = value_len.to_bytes_be()[31] as usize;
+                    let mut value = Vec::new();
+                    for _ in 0..value_len {
+                        if let Some(felt) = iter.next() {
+                            value.push(*felt);
+                        }
+                    }
+                    entries.push((key, value));
+                }
+            }
+        }
+
+        Ok(Self { entries })
+    }
+}
+
+impl From<(Vec<Felt>, Vec<Felt>)> for ContextStorageEntry {
+    fn from((key_felts, value_felts): (Vec<Felt>, Vec<Felt>)) -> Self {
+        let key = key_felts
+            .iter()
+            .flat_map(|f| f.to_bytes_be())
+            .filter(|&b| b != 0)
+            .collect();
+
+        let value = value_felts
+            .iter()
+            .flat_map(|f| f.to_bytes_be())
+            .filter(|&b| b != 0)
+            .collect();
+
+        ContextStorageEntry { key, value }
+    }
+}
+
+struct ValueCodec<'a>(&'a serde_json::Value);
+
+impl<'a> Encode for ValueCodec<'a> {
+    fn encode<W: FeltWriter>(&self, writer: &mut W) -> Result<(), Error> {
+        match self.0 {
+            serde_json::Value::Bool(b) => {
+                writer.write(Felt::from(*b as u64));
+                Ok(())
+            }
+            serde_json::Value::String(s) => {
+                if s.starts_with("0x") {
+                    // Handle hex string directly as a single Felt
+                    writer.write(
+                        Felt::from_hex(&s)
+                            .map_err(|e| Error::custom(&format!("Invalid hex string: {}", e)))?,
+                    );
+                } else {
+                    // Regular string - split into chunks
+                    let chunk_size = 31;
+                    let chunks: Vec<_> = s.as_bytes().chunks(chunk_size).collect();
+
+                    // Write number of chunks first
+                    writer.write(Felt::from(chunks.len()));
+
+                    // Write each chunk as a Felt
+                    for chunk in chunks {
+                        writer.write(Felt::from_bytes_be_slice(chunk));
+                    }
+                }
```

Comment:
just in case we have a string that starts with 0x and isn't a felt

```suggestion
                if s.starts_with("0x") {
                    // Attempt to handle hex string directly as a single Felt
                    if let Ok(felt) = Felt::from_hex(&s) {
                        writer.write(felt);
                        return Ok(());
                    }
                }

                // Regular string - split into chunks
                let chunk_size = 31;
                let chunks: Vec<_> = s.as_bytes().chunks(chunk_size).collect();

                // Write number of chunks first
                writer.write(Felt::from(chunks.len()));

                // Write each chunk as a Felt
                for chunk in chunks {
                    writer.write(Felt::from_bytes_be_slice(chunk));
                }
```